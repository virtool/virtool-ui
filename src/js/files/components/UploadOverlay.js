import React from "react";
import styled from "styled-components";
import { map, reject, sortBy } from "lodash-es";
import { connect } from "react-redux";
import { Badge, BoxGroup, BoxGroupHeader, BoxGroupSection } from "../../base";
import { UploadItem } from "./UploadItem";

const StyledUploadOverlay = styled.div`
    bottom: 0;
    ${props => (props.show ? "" : "display: none;")};
    max-width: 500px;
    padding: 0 15px 15px 0;
    position: fixed;
    right: 0;
    width: 35%;
    z-index: 90;
`;

const UploadOverlayContent = styled(BoxGroup)`
    background-color: ${props => props.theme.color.white};
    box-shadow: ${props => props.theme.boxShadow.lg};
    margin: 0;

    ${BoxGroupHeader} {
        display: block;
    }
`;

const UploadOverlayList = styled(BoxGroupSection)`
    height: auto;
    max-height: 200px;
    overflow-x: hidden;
    padding: 0;
`;

export const UploadOverlay = ({ uploads }) => {
    if (uploads.length) {
        const uploadComponents = map(uploads, upload => <UploadItem key={upload.localId} {...upload} />);

        return (
            <StyledUploadOverlay show={uploads.length}>
                <UploadOverlayContent>
                    <BoxGroupHeader>
                        <strong>Uploads</strong> <Badge>{uploadComponents.length}</Badge>
                    </BoxGroupHeader>
                    <UploadOverlayList>{uploadComponents}</UploadOverlayList>
                </UploadOverlayContent>
            </StyledUploadOverlay>
        );
    }

    return null;
};

export const mapStateToProps = state => {
    return { uploads: sortBy(reject(state.files.uploads, { fileType: "reference" }), "progress") };
};

export default connect(mapStateToProps)(UploadOverlay);
