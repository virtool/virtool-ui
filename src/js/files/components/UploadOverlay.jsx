import { formatDuration, intervalToDuration } from "date-fns";
import { every, map, reject, reverse, sumBy } from "lodash-es";
import React from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { getFontSize, getFontWeight } from "../../app/theme";
import { Badge, BoxGroup, BoxGroupHeader, BoxGroupSection } from "../../base";
import { UploadItem } from "./UploadItem";

const StyledUploadOverlay = styled.div`
    bottom: 0;
    ${props => (props.show ? "" : "display: none;")};
    max-width: 500px;
    padding: 0 15px 15px 0;
    position: fixed;
    right: 0;
    width: 35%;
    z-index: 90;
`;

const UploadOverlayContent = styled(BoxGroup)`
    background-color: ${props => props.theme.color.white};
    box-shadow: ${props => props.theme.boxShadow.lg};
    margin: 0;

    ${BoxGroupHeader} {
        align-items: auto;
        display: block;
        font-weight: ${getFontWeight("thick")};
        font-size: ${getFontSize("lg")};
    }
`;

const UploadOverlayList = styled(BoxGroupSection)`
    height: auto;
    max-height: 200px;
    overflow-x: hidden;
    padding: 0;
`;

const StyledUploadInformation = styled.div`
    display: flex;
    justify-content: space-between;
    font-size: ${getFontSize("md")};
    font-weight: ${getFontWeight("normal")};
    margin-top: 5px;
`;

const getFormat = timeInterval => {
    if (timeInterval.hours) {
        return "hours";
    } else if (timeInterval.minutes) {
        return "minutes";
    } else if (timeInterval.seconds) {
        return "seconds";
    }
    return "";
};

const UploadTiming = ({ remaining, uploadSpeed }) => {
    const remainingTime = remaining ? remaining : 3600;
    const estimatedUploadSpeed = uploadSpeed ? (uploadSpeed / 1000000).toFixed(0) : 0;
    const timeRemainingInterval = intervalToDuration({ start: 0, end: remainingTime * 1000 });

    let formattedTimeRemaining = "";
    if (timeRemainingInterval.hours > 12) {
        formattedTimeRemaining = " > 12hr remaining";
    } else {
        formattedTimeRemaining = formatDuration(
            {
                hours: timeRemainingInterval.hours,
                minutes: timeRemainingInterval.minutes,
                seconds: timeRemainingInterval.seconds,
            },

            {
                format: [getFormat(timeRemainingInterval)],
            },
        );
        formattedTimeRemaining = formattedTimeRemaining.length
            ? `${formattedTimeRemaining} remaining`
            : "0 seconds remaining";
    }

    return (
        <StyledUploadInformation>
            <div>{formattedTimeRemaining}</div>
            <div>{`${estimatedUploadSpeed} MB/s`}</div>
        </StyledUploadInformation>
    );
};

export const UploadOverlay = ({ uploads }) => {
    if (uploads.length) {
        const totalRemainingUploadTime = sumBy(uploads, "remaining");
        const overallUploadSpeed = sumBy(uploads, "uploadSpeed");
        const uploadComponents = map(uploads, upload => <UploadItem key={upload.localId} {...upload} />);

        const allUploadsComplete = every(uploads, ["progress", 100]);

        return (
            <StyledUploadOverlay show={uploads.length}>
                <UploadOverlayContent>
                    <BoxGroupHeader>
                        Uploads <Badge>{uploadComponents.length}</Badge>
                        {allUploadsComplete ? (
                            <StyledUploadInformation>Finishing uploads</StyledUploadInformation>
                        ) : (
                            <UploadTiming remaining={totalRemainingUploadTime} uploadSpeed={overallUploadSpeed} />
                        )}
                    </BoxGroupHeader>
                    <UploadOverlayList>{uploadComponents}</UploadOverlayList>
                </UploadOverlayContent>
            </StyledUploadOverlay>
        );
    }

    return null;
};

export const mapStateToProps = state => {
    return { uploads: reverse(reject(state.files.uploads, { type: "reference" })) };
};

export default connect(mapStateToProps)(UploadOverlay);
