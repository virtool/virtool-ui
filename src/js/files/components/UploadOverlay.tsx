import UploaderDialog from "@files/components/UploaderDialog";
import { useUploaderStore } from "@files/uploader";
import { cn } from "@utils/utils";
import React from "react";

/**
 * Overlay uploads with their progress and speeds.
 */
export default function UploadOverlay(): JSX.Element | null {
    const uploads = useUploaderStore(state => state.uploads);
    const remaining = useUploaderStore(state => state.remaining);
    const speed = useUploaderStore(state => state.speed);

    if (uploads.length === 0) {
        return null;
    }

    return (
        <div className={cn("fixed bottom-0 right-0 w-96 pr-4 pb-4 z-50")}>
            <UploaderDialog remaining={remaining} speed={speed} uploads={uploads} />
        </div>
    );
}
