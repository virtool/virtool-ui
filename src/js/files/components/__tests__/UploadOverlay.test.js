import { UploadOverlay, mapStateToProps } from "../UploadOverlay";

describe("<UploadOverlay />", () => {
    let props;

    beforeEach(() => {
        props = {
            uploads: [
                {
                    fileType: "subtraction",
                    localId: "123abc",
                    name: "test_reads.fastq.gz",
                    progress: 100,
                    size: 1024
                },
                {
                    fileType: "reads",
                    localId: "456def",
                    name: "test_reads.fastq.gz",
                    progress: 0,
                    size: 1024
                },
                {
                    fileType: "reads",
                    localId: "789ghi",
                    name: "test_reads.fastq.gz",
                    progress: 50,
                    size: 1024
                }
            ]
        };
    });

    it("should render", () => {
        const wrapper = shallow(<UploadOverlay {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should return null if no uploads", () => {
        props.uploads = [];
        const wrapper = shallow(<UploadOverlay {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("mapStateToProps", () => {
    let state;

    beforeEach(() => {
        state = {
            files: {
                uploads: [
                    { localId: "foo", progress: 0 },
                    { localId: "bar", progress: 12 },
                    { localId: "baz", progress: 37 }
                ]
            }
        };
    });

    it("should return sorted uploads in props", () => {
        const props = mapStateToProps(state);
        const [baz, bar, foo] = props.uploads;

        expect(props).toEqual({
            uploads: [baz, bar, foo]
        });
    });

    it("should exclude reference uploads", () => {
        state.files.uploads[1].fileType = "reference";
        const props = mapStateToProps(state);
        const [baz, foo] = props.uploads;
        expect(props).toEqual({
            uploads: [baz, foo]
        });
    });
});
