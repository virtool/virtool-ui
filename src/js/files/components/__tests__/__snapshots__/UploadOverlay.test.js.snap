// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UploadOverlay /> should render 1`] = `
<UploadOverlay__StyledUploadOverlay
  show={3}
>
  <UploadOverlay__UploadOverlayContent>
    <Box__BoxGroupHeader>
      <strong>
        Uploads
      </strong>
       
      <Badge>
        3
      </Badge>
    </Box__BoxGroupHeader>
    <UploadOverlay__UploadOverlayList>
      <UploadItem
        fileType="subtraction"
        key="123abc"
        localId="123abc"
        name="test_reads.fastq.gz"
        progress={100}
        size={1024}
      />
      <UploadItem
        fileType="reads"
        key="456def"
        localId="456def"
        name="test_reads.fastq.gz"
        progress={0}
        size={1024}
      />
      <UploadItem
        fileType="reads"
        key="789ghi"
        localId="789ghi"
        name="test_reads.fastq.gz"
        progress={50}
        size={1024}
      />
    </UploadOverlay__UploadOverlayList>
  </UploadOverlay__UploadOverlayContent>
</UploadOverlay__StyledUploadOverlay>
`;

exports[`<UploadOverlay /> should return null if no uploads 1`] = `""`;
